libnavigation-base/src/main/java/com/mapbox/navigation/base/geodeeplink/GeoDeeplinkParser.kt

```diff
@@ -0,0 +1,100 @@
+package com.mapbox.navigation.base.geodeeplink
+
+import com.mapbox.geojson.Point
+import java.net.URLDecoder
+
+/**
+ * Converts external geo intents into [GeoDeeplink] objects.
+ *
+ * Public documentation for the geo deeplink can be found here
+ * https://developers.google.com/maps/documentation/urls/android-intents
+ *
+ * Variations supported.
+ *
+ * - Coordinates without a place query
+ *     geo:37.757527,-122.392937
+ * - Coordinates with a place query
+ *     geo:37.788151,-122.407543?q=3107 Washington Street, San Francisco, California 94115, United States
+ * - Place query without coordinates
+ *     geo:0,0?q=3107 Washington Street, San Francisco, California 94115, United States
+ * - Encoded deep-links
+ *     geo:0,0?q=%E5%93%81%E5%B7%9D%E5%8C%BA%E5%A4%A7%E4%BA%95%206-16-16%20%E3%83%A1%E3%82%BE%E3%83%B3%E9%B9%BF%E5%B3%B6%E3%81%AE%E7%A2%A7201%4035.595404%2C139.731737
+ * - AT(@) symbol specifying coordinates
+ *     geo:0,0?q=Coffee Shop@37.757527,-122.392937
+ * - Parenthesis specifying the place
+ *     geo:0,0?q=54.356152,18.642736(ul. 3 maja 12, 80-802 Gdansk, Poland)
+ */
+object GeoDeeplinkParser {
+
+    /**
+     * Convert a string into a [GeoDeeplink] object.
+     * Returns a [GeoDeeplink] when the string starts with "geo:"
+     *
+     * [GeoDeeplink.point] or [GeoDeeplink.placeQuery] will provide
+     * a supported value or else the [GeoDeeplink] will be null.
+     */
+    fun parse(geoDeeplink: String?): GeoDeeplink? {
+        return if (geoDeeplink != null && geoDeeplink.startsWith("geo:")) {
+            val query = geoDeeplink.substring("geo:".length)
+            val args = query.split("?")
+            val point = args[0].toPoint() ?: args.query()?.queryCoordinates()
+            val placeQuery = args.query()?.removeCoordinates()
+            return if (point == null && placeQuery.isNullOrEmpty()) {
+                null
+            } else {
+                GeoDeeplink(
+                    deeplink = geoDeeplink,
+                    point = point,
+                    placeQuery = placeQuery
+                )
+            }
+        } else {
+            null
+        }
+    }
+
+    private fun String.toPoint(): Point? {
+        val coordinates = this.split(",")
+        if (coordinates.size < 2) return null
+        val latitude = coordinates[0].toCoordinate()
+        val longitude = coordinates[1].toCoordinate()
+            ?: coordinates[1].replace("%20", "").toCoordinate()
+        return if (latitude != null && longitude != null && (latitude != 0.0 || longitude != 0.0)) {
```

Comment:
why `0.0` `lat` and `lng` are filtered? ther're valid coordinates.

Reply:
`geo:0,0` is often used to specify a "null" location. See the `GeoDeeplinkParserTest` for examples like this these
```
"geo:0,0?q=54.356152,18.642736(ul. 3 maja 12, 80-802 Gdansk, Poland)",
"geo:0,0?q=1600 Amphitheatre Parkway, Mountain+View, California",
```

This is why it is an `OR` though, to ensure this deep link continues to work for every 0 location besides the 0,0. 
```
"geo:37.788151,0.0",
```

If you want to navigate your car to `0,0`, you can navigate a couple inches away `0.0000001,0.0` :)

Reply:
I agree, it's not valid coordinates 😄 IMO parer is not that place where we should have a logic right/wrong coordinates filtering

Reply:
> IMO parer is not that place where we should have a logic right/wrong coordinates filtering

I'm not sure what you mean. Are you saying the parser should treat 0,0 as a valid coordinate in this case?
 `"geo:0,0?q=1600 Amphitheatre Parkway, Mountain+View, California",`

Or are we in agreement, that anything that sends that deep link does-not intend to navigate to 0,0 coordinates?

Reply:
first one. If anyone sends a deep link with null-island coordinates we should handle it and try to parse and only then show an error "null island cannot be handled" or try to assemble route to null-island and the router trow error "cannot find route". But currently, we silently skip that deep-link and don't provide feedback what is wrong with that link

Reply:
> first one. If anyone sends a deep link with null-island coordinates we should handle it and try to parse and only then show an error "null island cannot be handled" or try to assemble route to null-island and the router trow error "cannot find route". But currently, we silently skip that deep-link and don't provide feedback what is wrong with that link

But that won't work for the features we're building. For example, when you navigate with audio input by speaking "navigate to a coffee shop"; the deep-link will be of form "geo:0,0?q=Coffee". The coordinates are unknown at this time, and the coordinates are required in the deep link.

How do you handle a geo deeplink when the coordinates are unknown @RingerJK ?

Reply:
I read `GeoDeeplinkParser` kdoc, and it makes sense for me now, we accept 0,0 as not-provided coords 👍 resolve the conversation